src/cpp/rtps/history/TopicPayloadPool.cpp

```diff
@@ -0,0 +1,282 @@
+// Copyright 2020 Proyectos y Sistemas de Mantenimiento SL (eProsima).
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+/**
+ * @file TopicPayloadPool.cpp
+ */
+
+#include <rtps/history/ITopicPayloadPool.h>
+
+#include "./TopicPayloadPool.hpp"
+#include "./TopicPayloadPool_impl/Preallocated.hpp"
+#include "./TopicPayloadPool_impl/PreallocatedWithRealloc.hpp"
+#include "./TopicPayloadPool_impl/Dynamic.hpp"
+#include "./TopicPayloadPool_impl/DynamicReusable.hpp"
+
+#include <memory>
+
+namespace eprosima {
+namespace fastrtps {
+namespace rtps {
+
+bool TopicPayloadPool::get_payload(
+        uint32_t size,
+        CacheChange_t& cache_change)
+{
+    PayloadNode* payload = nullptr;
+
+    std::lock_guard<std::mutex> lock(mutex_);
+    if (free_payloads_.empty())
+    {
+        payload = allocate(size); //Allocates a single payload
+        if (payload == nullptr)
+        {
+            cache_change.serializedPayload.data = nullptr;
+            cache_change.serializedPayload.max_size = 0;
+            cache_change.payload_owner(nullptr);
+            return false;
+        }
+    }
+    else
+    {
+        payload = free_payloads_.back();
+        free_payloads_.pop_back();
+    }
+
+    payload->reference();
+    cache_change.serializedPayload.data = payload->data();
+    cache_change.serializedPayload.max_size = payload->data_size();
+    cache_change.payload_owner(this);
+    return true;
+}
+
+bool TopicPayloadPool::get_payload(
+        SerializedPayload_t& data,
+        IPayloadPool*& data_owner,
+        CacheChange_t& cache_change)
+{
+    assert(cache_change.writerGUID != GUID_t::unknown());
+    assert(cache_change.sequenceNumber != SequenceNumber_t::unknown());
+
+    if (data_owner == this)
+    {
+        std::lock_guard<std::mutex> lock(mutex_);
+        PayloadNode* payload = all_payloads_.at(PayloadNode::data_index(data.data));
+        payload->reference();
```

Comment:
Could we make the reference counting without taking the mutex?

Reply:
We can make the reference counting without the mutex, but we cannot get the `PayloadNode` pointer from `all_payloads` since the index field can be modified if some history is released by another endpoint.

I created static methods to handle the size and counter fields from the buffer pointer instead of having to retrieve the `PayloadNode` from `all_payloads`